ember_debug/object-inspector.js

```diff
@@ -391,12 +549,31 @@ export default EmberObject.extend(PortMixin, {
     let mixins = Mixin.mixins(object);
     let mixinDetails = [];
 
-    let ownProps = propertiesForMixin({ mixins: [{ properties: object }] });
-    mixinDetails.push({ name: 'Own Properties', properties: ownProps, expand: true });
-
-    if (!Object.getPrototypeOf(object).hasOwnProperty('_super')) {
-      mixins.splice(0, 0, this.es6ClassMixinForObject(object));
+    let objectMixins = this.mixinDetailsForObject(object);
+    // merge first 2 into own properties, if its not created directly from EmberObject.create(...)
+    if (objectMixins.length > 1 && !objectMixins[1].isEmberObject) {
+      const allProperties = Ember.A(objectMixins[0].properties.concat(...objectMixins[1].properties));
+      objectMixins[0].properties = allProperties.uniqBy('name').toArray();
+      objectMixins.splice(1, 1);
+    }
+    // find first ember extended object to show mixins afterwards
+    let index = objectMixins.findIndex(o => o.isEmberExtended);
+    if (index < 0) index = objectMixins.length;
+    if (index <= 1) {
+      index = 1;
+      // remove properties that are in mixins, but not itself
+      if (mixins[0] && mixins[0].mixins) {
+        mixins[0].mixins.slice(0, -1).forEach((m) => {
+          const props = Object.keys((m.mixins && m.mixins[0] && m.mixins[0].properties) || {});
+          objectMixins[0].properties = objectMixins[0].properties.filter(o => !props.includes(o.name));
+        });
+      }
+      // the base mixins appear again later in the mixins...
+      mixins.splice(0, 1);
     }
+    mixinDetails.push(...objectMixins.slice(0, index));
+    const objectName = mixinDetails[0].name;
+
     mixins.forEach(mixin => {
```

Comment:
So, the new strategy appears to be:

1. Walk the entire prototype tree, gathering all information about the object/classes and putting it onto some fake-mixins
2. Put these on top, and then add mixins, and then add `EmberObject` at the very bottom

I think this is a bit problematic, because then the mixins are going to be out of order with the class prototypes they actually belong to. They'll also consist entirely of properties/descriptors that already show up on one of the prototype-mixins (unless they're overridden, which I guess technically they would all be?), e.g.:


```js
class Foo extends Bar.extend(MyMixin) {}

class Bar extends Baz {}

class Baz extends EmberObject {}
```
```
// inspector shows
- Foo 
- Bar
- Baz
- MyMixin
- EmberObject
```

I think ideally what we _should_ do here is show only the classes themselves by default, and then have a sort of subview where you can examine the mixins of a particular class:

```
// inspector shows
- Foo
- Bar
  - mixins: MyMixin
- Baz
- EmberObject
```

I think we could add the subview in a followup, but I'm not sure if we can show the out of order view in the meantime, that seems like it'd be really confusing. Maybe we could try to setup the order the correct way at least?

```
// inspector shows
- Foo
- Bar
- MyMixin
- Baz
- EmberObject
```

Reply:
I changed to logic for this part and added a test to check the correct order of prototypes and mixins